### ISSUE TITLE
Ranger extremely slow when browsing (using vim j-k movements) in certain FUSE mounted environments.

### ISSUE THREAD

**Issue by nicholasdewaal** [ranger/ranger#1025]
#### Runtime Environment

- Operating system and version: 16.04.2-Ubuntu
- Terminal emulator and version: Konsole 15.12.3
- Python version: 2.7.12
- Ranger version/commit: ranger 1.9.0b6
- Locale: 

#### Current Behavior
I use Cryfs (https://www.cryfs.org/#) for encrypted cloud storage, and mount it using a FUSE file system.  All the data is synced/stored on a local NVMe SSD (fast).

When I browse through the files and folders using the (j-k key press) movements, ranger takes a significant amount of time between each item movement when I am in the FUSE mounted folders.   It makes browsing using Ranger very inconvenient.   I don't notice this problem when using vifm to browse the files.

#### Expected Behavior

Better speed or customization for movements in slowly processing FUSE mounted folders.

#### Context

 It makes browsing using Ranger very slow and inconvenient. 

#### Possible Solutions
Buffering previews?  The ability to disable recomputing previews/other on each move?

#### Steps to reproduce
1. Install Cryfs: https://www.cryfs.org/#
2. Mount a cryfs folder on a fast SSD or Hard drive:  cryfs basedir mountdir
3. Fill mountdir with lot of folders/data
4. Browse files in mountdir with ranger, notice how slow it is

My rc.conf:

set show_hidden false
#map <C-n> console touch
map <C-f> console touch 
map <C-r> console mkdir 
map <a> console rename
#map <C-d> console mkdir
map <C-d> quit
map <C-h> set show_hidden true
map <C-j> set show_hidden false
set vcs_aware false
map zg set vcs_aware true
map <BACKSPACE> cut
set vcs_backend_git local
set max_history_size 20
#ext jpg = gthumb "$@"
#set preview_images true
set preview_images false
set use_preview_script true
set preview_script ~/.config/ranger/scope.sh

#### Traceback

```

```

**Comment by vfreefly** [comment]
Git support is enabled when you notice this problem?
Git repos can be a REALLY slow with sshfs systems. So if ranger trying to scan a remote git repo - it will be really slow. 
And that is not a ranger problem.

**Comment by sammoth** [comment]
Does it stop if you disable previews? This is presumably because previews are generated and displayed synchronously in ranger and you can only scroll as fast as the previews are shown.

**Comment by toonn** [comment]
Another factor that'll slow down things like network filesystems is ranger's tendency to do a lot of stat calls.

But yes, in this case I suspect file, not just image, previews so try disabling those. You can do this only locally for the fuse-fs directories with either `setintag` and tagging all of them with a special tag or `setlocal` for each of the paths.

**Comment by toonn** [comment]
Oh, one other thing I noticed the report might be confused about ranger's config files. The `ext jpg = gthumb "$@"` has no business in `rc.conf`, it'd go in `rifle.conf`.

**Comment by toonn** [comment]
And another thing, the `map console something` lines make no sense whatsoever. Maybe you think it makes the `:touch` et al commands call to the shell builtins or something? That's not what console does. And this is not how map works, you need to provide a key sequence first so all of those maps map the sequence `console` to one of ranger's `:commands`. There's others that are even weirder like `map quit`, `map cut`.

I suggest you completely delete your configuration and run `ranger --copy-config=rc` then edit that, that should give you more of a chance to create a proper `rc.conf`.

**Comment by nicholasdewaal** [comment]
I disabled previews and it is still slow in a FUSE mounted folder. I am
using ranger-stable 1.7.1

**Comment by nicholasdewaal** [comment]
Thank you,

I deleted my config files for ranger, and set up a fresh environment for
ranger using the command you suggested. I then disabled previews.
Unfortunately in my FUSE environment, it is still extremely slow. It is
fast however when I use vifm.

I am using ranger-stable 1.7.1.

Thanks,
Nick

**Comment by toonn** [comment]
I suspect vifm does way less than ranger does. Does vifm display the sizes of everything in the network filesystem for example? Does it update them when some file on the network filesystem changes? Ranger probably does too many stat calls, like I said. It's possible vifm does too few though.

I'd definitely like to optimize ranger's interaction with fuse and nfs filesystems. We just don't have much time to look into these things currently.

**Comment by nicholasdewaal** [comment]
Agreed, ranger is much better than vifm, even with ranger being slow with
FUSE. I just thought it might provide insight into what's causing the
slowdown by comparing differences. As you say, it's probably the stats.

If you do ever get the time, maybe it would be better to have stats and
previews simply fail to display while navigating unless that information is
available or cached? The slowdown happens on every j or k movement. Some
kind of processing goes on between each jump that slows it down. It would
be interesting to profile it.

Thanks again,
Nick

**Comment by Vifon** [comment]
@nicholasdewaal, you may find the "freeze" functionality handy. Press `F` (capital!) and ranger shouldn't read any more directories until you explicitly unfreeze it. I'm not sure if it's available for ranger 1.7.1 though. In general, disabling the previews (for both files and directories) greatly improves the performance with FUSE.

**Comment by nicholasdewaal** [comment]
As a side note, I noticed that j-k movements in ranger are fast in a FUSE
environment after at least one file or folder is selected (using the space
bar). That tells me that it's not previews or data loading that slows it
down.

-Nick

**Comment by toonn** [comment]
Hmm, maybe it's something about the selection then? Not having anything marked would mean the selection changes with every movement. I wouldn't expect setting the selection to do much of anything though but maybe my intuition is off.

**Comment by totorigolo** [comment]
My home is mounted on AFS and ranger takes 5 sec to load. However, I found that setting `freeze_files` to `true` and only unfreezing when entering a new directory greatly helps. However, I have to unfreeze/freeze a lot :/ Also, doing so shows a loading bar instead of just "loading...".

Hope that helps :)

**Comment by Kalagon** [comment]
Just in case someone is interested in this:
I also experience severe lag on fuse file systems when using previews. In my case it seems to be mostly caused by previews. The small hiccups caused by size calculation and other things are fine for me.

Since I don't really _need_ previews on the paths this happens, I'm working around it by escaping from the `scope.sh` script early. It's not perfect but it works well enough in my case.

This is the code that should be put quite far up in the script:
```bash
# do not process files on fuse file systems as it slows down ranger considerably
if [ "$(stat --file-system --format=%T $FILE_PATH)" == "fuseblk" ]; then
    exit 0
fi
```

**Comment by toonn** [comment]
@Kalagon, you could also just use `:setinpath` or `:setintag` to disable previews locally, which we'd recommend instead.

**Comment by nicholasdewaal** [comment]
I figured out why ranger is so slow in a fuse environment.  I ran "python
-m cProfile -o ranger_test.cprof ranger/ranger.py" to get a profile, and
it's statvfs that uses all the resources.

get_free_space in statusbar.py is using massive amounts of resources.  I'm
not quite sure how to modify the code to only refresh the "free" part of
the status bar to only update when changing folders.

-Nick

**Comment by toonn** [comment]
That's what I said in my first comment on this issue. Ranger goes ham on stat calls. That could definitely be improved but then again it's usually only an issue on slow filesystems, i.e., fuse or network filesystems.

**Comment by eMPee584** [comment]
However i just figured out it goes wild on stat calls to git-IGNORED subtrees..

**Comment by toonn** [comment]
Version control repositories have little to do with FUSE and remote filesystems though. Keep discussion to appropriate issues please.

**Comment by chase-moskal** [comment]
after using `s3fs-fuse` to mount an s3 bucket, i found that ranger felt unpleasantly slow there.

i used this command to disable ranger's preview feature (now it seems more responsive)

```sh
echo "set preview_files false" >> ~/.config/ranger/rc.conf
```
